From 04528268aa1cc1b45b1d413b26623d55177bda05 Mon Sep 17 00:00:00 2001
From: Stephen Smalley <stephen.smalley.work@gmail.com>
Date: Wed, 2 Oct 2024 16:14:34 -0400
Subject: [PATCH 52/90] selinux: refactor selinux_state_create()

Refactor selinux_state_create() to be more like create_user_ns()
after which it was originally modeled. In particular, pass in
a single cred argument and update the cred SELinux blob with
the new state. This makes the reference counting situation
clearer with regard to the old state / parent reference and
simplifies the callers.

Signed-off-by: Stephen Smalley <stephen.smalley.work@gmail.com>

Upstream-commit: 6f8c67b1ee8d55585f1f37f8024e06c303fbbf93

Backport: retain the 6.18 VM_DATA_DEFAULT_FLAGS API while initializing the initial namespace credentials.
---
 security/selinux/hooks.c            | 80 ++++++++++++++++++++---------
 security/selinux/include/security.h |  3 +-
 2 files changed, 57 insertions(+), 26 deletions(-)

diff --git a/security/selinux/hooks.c b/security/selinux/hooks.c
index 9d324dd1ad2b0a1c0b3ca74b67a8905329980f9a..6835b08c764ba31710527558ab351e86e594e0dc 100644
--- a/security/selinux/hooks.c
+++ b/security/selinux/hooks.c
@@ -203,22 +203,6 @@ static int selinux_lsm_notifier_avc_callback(u32 event)
 	return 0;
 }
 
-struct selinux_state *init_selinux_state;
-
-/*
- * initialise the security for the init task
- */
-static void cred_init_security(void)
-{
-	struct cred_security_struct *crsec;
-
-	/* NOTE: the lsm framework zeros out the buffer on allocation */
-
-	crsec = selinux_cred(unrcu_pointer(current->real_cred));
-	crsec->osid = crsec->sid = SECINITSID_KERNEL;
-	crsec->state = get_selinux_state(init_selinux_state);
-}
-
 /*
  * get the security ID of a set of credentials
  */
@@ -7902,9 +7886,10 @@ static struct security_hook_list selinux_hooks[] __ro_after_init = {
 
 static void selinux_state_free(struct work_struct *work);
 
-int selinux_state_create(struct selinux_state *parent,
-			 struct selinux_state **state)
+int selinux_state_create(const struct cred *cred)
 {
+	struct cred_security_struct *crsec = selinux_cred(cred);
+	struct selinux_state *parent = crsec->state;
 	struct selinux_state *newstate;
 	int rc;
 
@@ -7922,10 +7907,38 @@ int selinux_state_create(struct selinux_state *parent,
 	if (rc)
 		goto err;
 
+	if (parent) {
+		/*
+		 * The reference to the new state replaces the reference
+		 * to the old state (parent) in the cred security blob;
+		 * hence, we do not need to use get_selinux_state() below
+		 * to increment the parent reference count.
+		 */
+		newstate->parent = parent;
+	}
+
+	/*
+	 * Set the new namespace.
+	 * The reference count was initialized to 1 and
+	 * this is that reference.
+	 */
+	crsec->state = newstate;
+
+	/* Reset the SIDs for the new namespace. */
 	if (parent)
-		newstate->parent = get_selinux_state(parent);
+		crsec->osid = crsec->sid = SECINITSID_INIT;
+	crsec->exec_sid = crsec->create_sid = crsec->keycreate_sid =
+		crsec->sockcreate_sid = SECSID_NULL;
+
+	/*
+	 * Save the credential in the parent namespace
+	 * for later use in checks in that namespace.
+	 */
+	if (parent) {
+		put_cred(crsec->parent_cred);
+		crsec->parent_cred = get_current_cred();
+	}
 
-	*state = newstate;
 	return 0;
 err:
 	kfree(newstate);
@@ -7953,16 +7966,35 @@ void __put_selinux_state(struct selinux_state *state)
 	schedule_work(&state->work);
 }
 
+struct selinux_state *init_selinux_state;
+
 static __init int selinux_init(void)
 {
+	const struct cred *cred = unrcu_pointer(current->real_cred);
+	struct cred_security_struct *crsec = selinux_cred(cred);
+
 	pr_info("SELinux:  Initializing.\n");
 
-	if (selinux_state_create(NULL, &init_selinux_state))
+	/*
+	 * Initialize the first cred with the kernel SID and
+	 * NULL state since selinux_state_create() expects
+	 * these two fields to be set. The rest is handled by
+	 * selinux_state_create(), which will update the state
+	 * field to refer to the new state and set the parent
+	 * pointer to the old state value (NULL).
+	 */
+	crsec->osid = crsec->sid = SECINITSID_KERNEL;
+	crsec->state = NULL;
+	if (selinux_state_create(cred))
 		panic("SELinux: Could not create initial namespace\n");
-	enforcing_set(init_selinux_state, selinux_enforcing_boot);
 
-	/* Set the security state for the initial task. */
-	cred_init_security();
+	/*
+	 * Save a reference to the initial SELinux namespace
+	 * for use in various other functions.
+	 */
+	init_selinux_state = get_selinux_state(crsec->state);
+
+	enforcing_set(init_selinux_state, selinux_enforcing_boot);
 
 	/* Inform the audit system that secctx is used */
 	audit_cfg_lsm(&selinux_lsmid,
diff --git a/security/selinux/include/security.h b/security/selinux/include/security.h
index 10ab314c6bf9b6051bd200bd89aba306f3c21c54..49dc184764dd2ad4d437514da3ba73ddb3dbcf73 100644
--- a/security/selinux/include/security.h
+++ b/security/selinux/include/security.h
@@ -113,8 +113,7 @@ struct selinux_state {
 	struct work_struct work;
 } __randomize_layout;
 
-int selinux_state_create(struct selinux_state *parent,
-			 struct selinux_state **state);
+int selinux_state_create(const struct cred *cred);
 void __put_selinux_state(struct selinux_state *state);
 
 void selinux_policy_free(struct selinux_policy *policy);
